Client/Frontend/Library/HistoryPanel/HistoryPanel.swift

```diff
@@ -0,0 +1,651 @@
+        
+        if let historyActionableCell = clearHistoryCell {
+            setTappableStateAndStyle(with: HistoryActionablesModel.activeActionables.first, on: historyActionableCell)
+        }
+        
+        applySnapshot(animatingDifferences: true)
+    }
+    
+    func siteAt(indexPath: IndexPath) -> Site? {
+        guard let siteItem = diffableDatasource?.itemIdentifier(for: indexPath) as? Site else { return nil }
+        
+        return siteItem
+    }
+
+    private func showClearRecentHistory() {
+        clearHistoryHelper.showClearRecentHistory(onViewController: self, didComplete: { [weak self] date in
+            if let date = date {
+                self?.viewModel.removeVisibleSectionFor(date: date)
+            } else {
+                // The only time there's no date, is when we are deleting everything.
+                self?.viewModel.visibleSections = []
+            }
+            
+            self?.viewModel.reloadData()
+            self?.applySnapshot(animatingDifferences: true)
+            
+            if let cell = self?.clearHistoryCell {
+                self?.setTappableStateAndStyle(with: HistoryActionablesModel.activeActionables.first(where: { $0.itemIdentity == .clearHistory }), on: cell)
+            }
+        })
+    }
+    
+    func siteForIndexPath(_ indexPath: IndexPath) -> Site? {
+        // First section is reserved for fixed history actionables.
+        guard indexPath.section > HistoryPanelSections.additionalHistoryActions.rawValue else {
+            return nil
+        }
+
+        let sitesInSection = viewModel.groupedSites.itemsForSection(indexPath.section - 1)
+        return sitesInSection[safe: indexPath.row]
+    }
+
+    func handleNotification(_ notification: Notification) {
+        switch notification.name {
+        case .FirefoxAccountChanged, .PrivateDataClearedHistory:
+            viewModel.groupedSites = DateGroupedTableData<Site>()
+            reloadData()
+
+            if profile.hasSyncableAccount() {
+                resyncHistory()
+            }
+            break
+        case .DynamicFontChanged:
+            reloadData()
+
+            if emptyStateOverlayView.superview != nil {
+                emptyStateOverlayView.removeFromSuperview()
+            }
+            emptyStateOverlayView = createEmptyStateOverlayView()
+            resyncHistory()
+            break
+        case .DatabaseWasReopened:
+            if let dbName = notification.object as? String, dbName == "browser.db" {
+                reloadData()
+            }
+        case .OpenClearRecentHistory:
+            showClearRecentHistory()
+        default:
+            // no need to do anything at all
+            browserLog.error("Error: Received unexpected notification \(notification.name)")
+            break
+        }
+    }
+
+    // MARK: - UITableViewDataSource
+    
+    /// Handles dequeuing the appropriate type of cell when needed.
+    private func configureDatasource() {
```

Comment:
This function in massive. Can it be broken down in littler bits?

Also, should these ifs be else ifs?

Reply:
I decided against else ifs b/c we're returning at each point anyways. What's your preference?

Reply:
I'm honestly not sure. I'm leaning towards else ifs simply because the context is clearer. However, this is in no way a blocker so it's totally fine.